src/query_ctrl.ts

```diff
@@ -1,26 +1,48 @@
 ///<reference path="../node_modules/grafana-sdk-mocks/app/headers/common.d.ts" />
 
 import { QueryCtrl } from 'app/plugins/sdk';
-import _ from 'lodash';
 import './css/query_editor.css!';
 
 export class VividCortexQueryCtrl extends QueryCtrl {
   static templateUrl = 'partials/query.editor.html';
 
-  private debouncedMetricFind;
+  public loading;
+
+  private metricFindDefer;
+  private metricFindTimeout;
+  private $q;
 
   /** @ngInject **/
-  constructor($scope, $injector) {
+  constructor($scope, $injector, $q) {
+    this.$q = $q;
+
     super($scope, $injector);
 
     this.target.target = this.target.target || 'select metric';
     this.target.type = this.target.type || 'timeseries';
-
-    this.debouncedMetricFind = _.debounce(this.datasource.metricFindQuery.bind(this.datasource), 250);
   }
 
   getOptions(query) {
-    return this.debouncedMetricFind(query);
+    if (!this.metricFindDefer) {
+      this.metricFindDefer = this.$q.defer();
+    }
+
+    this.loading = true;
+
+    clearTimeout(this.metricFindTimeout);
+
+    this.metricFindTimeout = setTimeout(() => {
```

Comment:
Shouldn't we use `$timeout` here instead, since we are in an angularJS context? And if yes, $timeout actually returns a promise which you can use in place of your `metricFindDefer`.

Reply:
Yep, that's true 🤦‍♂️ . Thanks for the suggestion.

Reply:
I still need to return the same promise and resolve it after the timeout, but I'm now using $timeout.